### PR TITLE
grpc update

### PR DESCRIPTION
**Pull request by michaeldjeffrey**
A PR that moves all the clients to use records will come next.

### CODE REVIEW

File: src/hpr_config_worker.erl [hunk 1]
```diff
@@ -87,59 +86,40 @@ init(#{host := Host, port := Port} = Args) ->
...
-            {noreply, State, {continue, ?CONNECT}};
-        {error, timeout} ->
-            lager:debug("rcv timeout"),
-            {_, Backoff1} = backoff:succeed(Backoff0),
-            {noreply, State#state{conn_backoff = Backoff1}, {continue, ?RCV_CFG_UPDATE}};
-        {error, E} ->
-            lager:error("failed to rcv ~p", [E]),
-            {stop, {error, E}}
+
+    %% NOTE: Client is still expecting maps.
```
Comment (macpie) [inline src/hpr_config_worker.erl:110]: I thought GRPC box expected records?

File: src/grpc/packet_router/hpr_packet_router_downlink_handler.erl [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+-export([
+    init/3,
+    handle_message/2,
+    handle_headers/2,
+    handle_trailers/4,
+    handle_eos/1
+]).
+
+-type stream_id() :: non_neg_integer().
+-type callback_data() :: #{gateway => libp2p_crypto:pubkey_bin(), stream_id => stream_id()}.
```
Comment (macpie) [inline src/grpc/packet_router/hpr_packet_router_downlink_handler.erl:14]: This has to be a `map()` right? We can't make this a record?

File: src/hpr_config_worker.erl [hunk 3]
```diff
@@ -87,59 +86,40 @@ init(#{host := Host, port := Port} = Args) ->
...
+    Input = hpr_route_stream_req:to_map(SignedRouteStreamReq),
+    Options = #{channel => ?CHANNEL_NAME},
+    case helium_config_route_client:stream(Input, Options) of
```
Comment (macpie) [inline src/hpr_config_worker.erl:113]: We used to do `send_last` here, anything equivalent? or `helium_config_route_client:stream` is enough for the server to know?

File: src/hpr_config_worker.erl (hunk 1 above)
Comment (michaeldjeffrey) [reply src/hpr_config_worker.erl:110]: For posterity:

We use [grpcbox_plugin](https://github.com/novalabsxyz/grpcbox_plugin).
Grpcbox can be configured to use maps _or_ records.

`grpc_client` _only_ supports maps.
So we configured our client modules to be generated using maps.

For the changover from `grpc_client` to `grpcbox_client`, only the client handlers are being swapped out.

File: src/grpc/packet_router/hpr_packet_router_downlink_handler.erl (hunk 2 above)
Comment (michaeldjeffrey) [reply src/grpc/packet_router/hpr_packet_router_downlink_handler.erl:14]: Yes, we can.

File: src/hpr_config_worker.erl (hunk 3 above)
Comment (michaeldjeffrey) [reply src/hpr_config_worker.erl:113]: `stream` is the name of the rpc.
We tell the generated client we want to use this RPC, and it gives us back a configured `stream` that we give to `grpcbox_client` to do so.

The code this is replacing is 
```erlang
{ok, Stream} = grpc_client:new_stream(
        Connection, 'helium.config.route', stream, client_config_pb
    ),
```

File: rebar.config [hunk 4]
```diff
@@ -13,13 +13,10 @@
     {gproc, "0.9.0"},
     {backoff, "1.1.6"},
     {clique, ".*", {git, "https://github.com/helium/clique.git", {branch, "develop"}}},
-    {grpcbox,
-        {git, "https://github.com/novalabsxyz/grpcbox.git",
-            {ref, "6180be0f4f4f199f824e60ae5c64ddd2daa5acff"}}},
+    {grpcbox, {git, "https://github.com/novalabsxyz/grpcbox.git", {branch, "mj/full-update"}}},
```
Comment (macpie) [inline rebar.config:16]: Can we get this PR merged?

File: src/grpc/config/hpr_cs_route_stream_worker.erl [hunk 5]
```diff
@@ -76,63 +74,42 @@ handle_cast(Msg, State) ->
...
-    case hpr_cs_conn_worker:get_connection() of
-        undefined ->
+    {PubKey, SigFun} = persistent_term:get(?HPR_KEY),
+    PubKeyBin = libp2p_crypto:pubkey_to_bin(PubKey),
+    RouteStreamReq = hpr_route_stream_req:new(PubKeyBin),
+    SignedRouteStreamReq = hpr_route_stream_req:sign(RouteStreamReq, SigFun),
+    SignedRouteStreamReqMap = hpr_route_stream_req:to_map(SignedRouteStreamReq),
+    StreamOptions = #{channel => config_channel},
+
+    case helium_config_route_client:stream(SignedRouteStreamReqMap, StreamOptions) of
```
Comment (macpie) [inline src/grpc/config/hpr_cs_route_stream_worker.erl:84]: This is much nicer, thanks!

File: src/grpc/config/hpr_cs_route_stream_worker.erl [hunk 6]
```diff
@@ -76,63 +74,42 @@ handle_cast(Msg, State) ->
...
+handle_info({data, _StreamID, RouteStreamRes}, #state{file_backup_path = Path} = State) ->
+    lager:debug("route update"),
+    ok = process_route_stream_res(hpr_route_stream_res:from_map(RouteStreamRes), Path),
+    {noreply, State};
 handle_info(
-    ?RCV_CFG_UPDATE,
-    #state{
-        stream = Stream, file_backup_path = Path, conn_backoff = Backoff0
-    } = State
+    {'DOWN', Ref, process, Pid, _Reason},
```
Comment (macpie) [inline src/grpc/config/hpr_cs_route_stream_worker.erl:101]: I don't see any monitor above, does helium_config_route_client:stream monitors by itself?

File: src/grpc/config/hpr_cs_skf_stream_worker.erl [hunk 7]
```diff
@@ -70,65 +68,42 @@ handle_cast(Msg, State) ->
...
+handle_info({data, _StreamID, SKFStreamRes}, State) ->
+    lager:debug("sfk update"),
+    ok = process_res(hpr_skf_stream_res:from_map(SKFStreamRes)),
+    {noreply, State};
 handle_info(
-    ?RCV_CFG_UPDATE,
-    #state{
-        stream = Stream, conn_backoff = Backoff0
-    } = State
+    {'DOWN', Ref, process, Pid, _Reason},
```
Comment (macpie) [inline src/grpc/config/hpr_cs_skf_stream_worker.erl:95]: I don't see any monitor above, does helium_config_session_key_filter_client:stream monitors by itself?

File: src/grpc/config/hpr_cs_skf_stream_worker.erl [hunk 8]
```diff
@@ -70,65 +68,42 @@ handle_cast(Msg, State) ->
...
-    lager:info("connection ~p went down ~p", [_Pid, _ExitReason]),
+    lager:warning("stream closed"),
     self() ! ?INIT_STREAM,
     {noreply, State#state{stream = undefined}};
+handle_info({headers, _StreamID, _Headers}, State) ->
+    %% noop on headers
+    {noreply, State};
+handle_info({trailers, _StreamID, _Trailers}, State) ->
+    %% noop on trailers
+    {noreply, State};
```
Comment (macpie) [inline src/grpc/config/hpr_cs_skf_stream_worker.erl:106]: Should we do any stream cleanup on terminate?

File: src/grpc/config/hpr_cs_sup.erl [hunk 9]
```diff
@@ -28,7 +28,6 @@ init([]) ->
 
     ConfigServiceConfig = application:get_env(?APP, config_service, #{}),
     ChildSpecs = [
-        ?WORKER(hpr_cs_conn_worker, [ConfigServiceConfig]),
```
Comment (macpie) [inline src/grpc/config/hpr_cs_sup.erl:31]: We can probably remove that SUP and move  back the stream worker to main SUP

File: src/hpr_sup.erl [hunk 10]
```diff
@@ -92,12 +93,7 @@ init([]) ->
         ?SUP(hpr_gwmp_sup, []),
 
         ?SUP(hpr_http_roaming_sup, []),
-        ?ELLI_WORKER(hpr_http_roaming_downlink_handler, [ElliConfigRoamingDownlink]),
-
-        ?WORKER(hpr_router_connection_manager, []),
-        ?WORKER(hpr_router_stream_manager, [
-            'helium.packet_router.packet', route, client_packet_router_pb
-        ])
+        ?ELLI_WORKER(hpr_http_roaming_downlink_handler, [ElliConfigRoamingDownlink])
```
Comment (macpie) [inline src/hpr_sup.erl:96]: We can probably remove `hpr_cs_sup` and move back the stream worker here

File: src/grpc/config/hpr_cs_route_stream_worker.erl (hunk 6 above)
Comment (michaeldjeffrey) [reply src/grpc/config/hpr_cs_route_stream_worker.erl:101]: https://github.com/novalabsxyz/grpcbox/blob/master/src/grpcbox_client_stream.erl#L50

The monitor is added when the stream is created.

File: src/grpc/config/hpr_cs_skf_stream_worker.erl (hunk 7 above)
Comment (michaeldjeffrey) [reply src/grpc/config/hpr_cs_skf_stream_worker.erl:95]: https://github.com/novalabsxyz/grpcbox/blob/master/src/grpcbox_client_stream.erl#L50

The monitor is added when the stream is created.

File: src/grpc/config/hpr_cs_skf_stream_worker.erl (hunk 8 above)
Comment (michaeldjeffrey) [reply src/grpc/config/hpr_cs_skf_stream_worker.erl:106]: Yes, looking into it right now.

Comment (michaeldjeffrey) [reply src/grpc/config/hpr_cs_skf_stream_worker.erl:106]: Stream Pids are linked to their Channels. 
When the channel goes down, it takes down all it's streams.

Workers like this one clean up by dumping the `grpcbox_stream:stream()` map.

In the protocol router, when a stream receives an EOS, it removes itself from ets.

File: rebar.config (hunk 4 above)
Comment (michaeldjeffrey) [reply rebar.config:16]: working on it. https://github.com/novalabsxyz/grpcbox/pull/16

File: src/grpc/config/hpr_cs_skf_stream_worker.erl [hunk 11]
```diff
@@ -70,65 +77,61 @@ handle_cast(Msg, State) ->
...
+handle_info({trailers, _StreamID, Trailers}, State) ->
+    %% IF a stream is closed by the server side, Trailers will be
+    %% received before the EOS. Removing the stream from state will
+    %% mean none of the other clauses match, and reconnecting will not
+    %% be attempted.
+    %% ref: https://grpc.github.io/grpc/core/md_doc_statuscodes.html
+    case Trailers of
+        {<<"12">>, _, _} ->
+            lager:notice(
```
Comment (macpie) [inline src/grpc/config/hpr_cs_skf_stream_worker.erl:119]: raise this to error?

Comment (michaeldjeffrey) [reply src/grpc/config/hpr_cs_skf_stream_worker.erl]: raised all the notices to errors.

File: src/grpc/config/hpr_cs_route_stream_worker.erl [hunk 12]
```diff
@@ -1,3 +1,48 @@
...
+%% Makes a GRPC stream to the config service to receive route updates
+%% and forward them dealt with.
+%%
+%% A "channel" is created `grpcbox' when the app is started. This
+%% channel does not make an actual connection. That happens when a
+%% stream is created and a message is sent.
+%%
+%% If a channel goes down, all the stream will receive a `eos'
+%% message, and a `DOWN' message. The channel will clean up the
+%% remaining stram pids.
```
Comment (macpie) [inline src/grpc/config/hpr_cs_route_stream_worker.erl:14]:
```suggestion
%% remaining stream pids.
```

File: src/grpc/config/hpr_cs_route_stream_worker.erl [hunk 13]
```diff
@@ -1,3 +1,48 @@
...
+%% In this care, we log the unimplimented message, but do not attempt
+%% to reconnect.
+%%
+%% = Undefined Channel =
+%%
+%% All workers that talk to the Config Service use the same client
+%% channel, `config_channel'. Channels do not make connections to
+%% servers. If this message is received it means the channel was never
+%% created, either through configuration, or explicitly with
+%% `grpcbox_clietn:connect/3'.
```
Comment (macpie) [inline src/grpc/config/hpr_cs_route_stream_worker.erl:36]:
```suggestion
%% `grpcbox_client:connect/3'.
```

File: src/grpc/config/hpr_cs_route_stream_worker.erl [hunk 14]
```diff
@@ -76,63 +119,61 @@ handle_cast(Msg, State) ->
...
-            SignedRouteStreamReq = hpr_route_stream_req:sign(RouteStreamReq, SigFun),
-            ok = grpc_client:send_last(Stream, hpr_route_stream_req:to_map(SignedRouteStreamReq)),
-            lager:info("stream initialized"),
-            self() ! ?RCV_CFG_UPDATE,
-            {noreply, State#state{stream = Stream, conn_backoff = Backoff1}}
+            {noreply, State#state{stream = undefined}};
+        _ ->
+            {noreply, State}
     end;
 handle_info(
```
Comment (macpie) [inline src/grpc/config/hpr_cs_route_stream_worker.erl:169]: Comment to explain that `eos` will act as `DOWN` message

File: src/grpc/packet_router/hpr_packet_router_downlink_handler.erl [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+]).
+
+-record(state, {
+    gateway :: libp2p_crypto:pubkey_bin(),
+    lns :: binary(),
+    stream_id :: stream_id()
+}).
+
+-type stream_id() :: non_neg_integer().
+-type callback_data() :: #state{}.
```
Comment (macpie) [inline src/grpc/packet_router/hpr_packet_router_downlink_handler.erl:20]:
```suggestion
-type state() :: #state{}.
```

**Review by macpie — APPROVED** [review]

**Pull request merged**